src/hdmf/common/hierarchicaltable.py

```diff
@@ -0,0 +1,258 @@
+    #          If this is the case than we need to de-normalize the data and flatten the hierarchy
+    if not hcol_target.has_foreign_columns():
+        # 1) Iterate over all rows in our hierarchical columns (i.e,. the DynamicTableRegion column)
+        #    First we here get a list of DataFrames, one for each row. If our hcol is a VectorIndex
+        #    (i.e., out column is a ragged array), then simply loading the data from the VectorIndex
+        #    will do the trick. If we have a regular VectorData column, then we need to load the
+        #    elements ourselves (using slice syntax to make sure we get DataFrames) one-row-at-a-time
+        if isinstance(hcol, VectorIndex):
+            rows = hcol.get(slice(None), index=False, df=True)
+        else:
+            rows = [hcol[i:(i+1)] for i in range(len(hcol))]
+        # Now we iterate over the rows, where each row is described by a DataFrame with one-or-more rows
+        for row_index, row_df in enumerate(rows):
+            # 1.1): Since hcol is a DynamicTableRegion, each row returns another pandas.DataFrame so we
+            #       next need to iterate over all rows in that table to denormalize our data
+            for row in row_df.itertuples(index=True):
+                # 1.1.1) Determine the column data for our row. Each selected row from our target table
+                #        becomes a row in our flattened table
+                data.append(row)
+                # 1.1.2) Determine the multi-index tuple for our row, consisting of: i) id of the row in this
+                #        table, ii) all columns (except the hierarchical column we are flattening), and
+                #        iii) the index (i.e., id) from our target row
+                index_data = ([dynamic_table.id[row_index], ] +
+                              [dynamic_table[row_index, colname]
+                               for colname in dynamic_table.colnames if colname != hcol_name])
+                # for i in indexed_column_indicies:  # Fix data from indexed columns
+                #    index_data[i] = tuple(index_data[i])  # Convert from list to tuple (which is hashable)
+                index.append(tuple(index_data))
+                # Determine the names for our index and columns of our output table if this is the first row.
+                # These are constant for all rows so we only need to do this onle once for the first row.
+                if row_index == 0:
+                    index_names = ([(dynamic_table.name, 'id')] +
+                                   [(dynamic_table.name, colname)
+                                    for colname in dynamic_table.colnames if colname != hcol_name])
+                    columns = pd.MultiIndex.from_tuples([(hcol_target.name, 'id'), ] +
+                                                        [(hcol_target.name, c) for c in row_df.columns],
+                                                        names=('source_table', 'label'))
```

Comment:
When would these lines result in a different `columns` than the `columns` generated by lines 98-100? i.e. can this be removed and the if condition on line 94 removed so that lines 95-100 are always run?

Reply:
In the "Case 2" part (i.e., in the else statement) the logic inside and outside of the loop is identical, so I was indeed able to removed the logic in the loop and updated the code per your suggestion.

In the "Case 1" part of the code, the key difference is that within the loop we can use ``row_df.columns`` to determine the columns and outside we have to look at ``hcol_target.colnames``. I need to take a closer look at why this leads to a difference (I honestly don't remember, as I wrote this part of the code nearly a year ago) but tests are failing without the logic, so I left this as is for now until I had a chance to investigate further.

Reply:
Ok, the culprit is ``AlignedDynamicTable``. In the case of a regular ``DynamicTable`` ``hcol_target.colnames`` will give us the full list of column names but for ``AlignedDynamicTable`` this is only the columns of the main table without the subtables. I.e., we should be able to eliminate the difference in logic here, but need to deal with this difference in how we determine the columns.